lib/full-site-editing/block-templates.php

```diff
@@ -415,20 +479,50 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
  * @return WP_Block_Template|null File template.
  */
 function gutenberg_get_block_file_template( $id, $template_type = 'wp_template' ) {
+	/**
+	 * Filters the block templates array before the query takes place.
+	 *
+	 * Return a non-null value to bypass the WordPress quries.
+	 *
+	 * @since 5.8
+	 *
+	 * @param WP_Block_Template|null $block_template Return block template object to short-circuit the default query,
+	 *                                               or null to allow WP to run it's normal queries.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
+	 */
+	$block_template = apply_filters( 'pre_get_block_file_template', null, $id, $template_type );
+	if ( ! is_null( $block_template ) ) {
+		return $block_template;
+	}
+
 	$parts = explode( '//', $id, 2 );
 	if ( count( $parts ) < 2 ) {
 		return null;
 	}
 	list( $theme, $slug ) = $parts;
 
-	if ( wp_get_theme()->get_stylesheet() === $theme ) {
-		$template_file = _gutenberg_get_template_file( $template_type, $slug );
-		if ( null !== $template_file ) {
-			return _gutenberg_build_template_result_from_file( $template_file, $template_type );
-		}
+	if ( wp_get_theme()->get_stylesheet() !== $theme ) {
+		return null;
 	}
 
-	return null;
+	$template_file = _gutenberg_get_template_file( $template_type, $slug );
+	if ( null === $template_file ) {
+		return null;
+	}
+
+	$block_template = _gutenberg_build_template_result_from_file( $template_file, $template_type );
```

Comment:
These early `return null`'s will bypass the filter you are adding at the end of the function, making it useless in these cases.

I wonder if it would be better to set `$block_template` as `null` in the case of unexpected theme or missing file, so that `null` return can instead be part of the final filter?

Reply:
Fair point, I've updated the returns to also call `apply_filters`